crates/acvm/src/pwg/binary.rs

```diff
@@ -0,0 +1,234 @@
+                return None;
+            }
+            max += BigUint::from_bytes_be(&c.to_bytes());
+        }
+        for (c, w) in &gate.linear_combinations {
+            if let Some(v) = self.get_max_value(w) {
+                max += BigUint::from_bytes_be(&c.to_bytes()) * v;
+            } else {
+                return None;
+            }
+        }
+        if max > FieldElement::modulus() {
+            return None;
+        }
+        Some(max + BigUint::from_bytes_be(&gate.q_c.to_bytes()))
+    }
+
+    fn solve_inverse(
+        &self,
+        gate: &Expression,
+        initial_witness: &mut BTreeMap<Witness, FieldElement>,
+    ) -> GateResolution {
+        if gate.mul_terms.len() == 1
+            && self.are_inverse(&gate.mul_terms[0].1, &gate.mul_terms[0].2)
+            && gate.linear_combinations.is_empty()
+        {
+            if gate.q_c.is_zero() {
+                initial_witness.insert(gate.mul_terms[0].1, FieldElement::zero());
+                initial_witness.insert(gate.mul_terms[0].2, FieldElement::zero());
+                return GateResolution::Resolved;
+            } else if !gate.q_c.is_one() {
+                return GateResolution::UnsatisfiedConstrain;
+            }
+        }
+
+        GateResolution::Skip
+    }
+
+    // look for boolean constraint and add boolean witness to a map
+    pub fn lookup_booleans(&mut self, arith: &Expression) {
+        let mut x = usize::MAX;
+        if arith.mul_terms.len() == 1 && arith.linear_combinations.len() == 1 {
+            // x*x = x
+            if arith.mul_terms[0].1 == arith.mul_terms[0].2
+                && arith.mul_terms[0].1 == arith.linear_combinations[0].1
+                && arith.q_c.is_zero()
+                && (arith.mul_terms[0].0 + arith.linear_combinations[0].0).is_zero()
+            {
+                x = 0;
+            } else {
+                // x = a*b, a,b booleans or inverse
+                if self.are_boolean(&arith.mul_terms[0].1, &arith.mul_terms[0].2) {
+                    if arith.q_c.is_zero() {
+                        if (arith.mul_terms[0].0 + arith.linear_combinations[0].0).is_zero() {
+                            x = 0;
+                        }
+                    } else if (arith.mul_terms[0].0 + arith.q_c).is_zero()
+                        && arith.mul_terms[0].0 == arith.linear_combinations[0].0
+                    {
+                        x = 0;
+                    }
+                }
+            }
+        } else if arith.mul_terms.is_empty() && arith.linear_combinations.len() == 2 {
+            //x=y
+            let z = if self.is_boolean(&arith.linear_combinations[0].1)
+                && !self.is_boolean(&arith.linear_combinations[1].1)
+            {
+                1
+            } else if self.is_boolean(&arith.linear_combinations[1].1)
+                && !self.is_boolean(&arith.linear_combinations[0].1)
+            {
+                0
+            } else {
+                usize::MAX
```

Comment:
Looks like there is a pervasive use of `usize::MAX` in this function as a default empty value followed by occasional checks if the given value is not this invalid value.

We should avoid these in general in favor of `Option<T>`, especially if we're performing arithmetic on them. It is more self-documenting, leaves less room for error, and makes it clearer to readers when we're checking for special cases. Similar the `< usize::MAX` or `!= usize::MAX` checks should both be `if let Some(var) = ...` which is more consistent.

Reply:
Done, I switched to Option.